docs/features/enhanced-using.md

```diff
@@ -0,0 +1,345 @@
+```c#
+    goto label1; // error CS8648: A goto cannot jump to a location after a using declaration.
+    using var x = ...;
+label1:
+    return;
+```
+
+```c#
+    goto label1; // ok. using declaration is in a lower scope than goto and label
+    {
+        using var x = ...;
+    }
+label1:
+    return;
+```
+
+```c#
+    {
+        {
+            goto label1; // error CS8648: A goto cannot jump to a location after a using declaration.          
+        }
+        using var x = ...;
+    }
+label1:
+    return;
+```
+
+It is forbidden to jump 'backwards' to a location before a using declaration, when the label is in the same scope as the using declaration. 
+```c#
+label1:
+        using var x = ...; //error CS8649: A goto cannot jump to a location before a using declaration within the same block.
+        goto label1;
+```
+
+Jumping 'backwards' to a label at a higher scope is specifically permitted.
+
+```c#
+label1:
+    {
+        using var x = ...;
+        goto label1; // allowed. label1 is in a higher scope than the using declaration
+    }
+```
+
+When jumping to a higher scope, any declared using variables will be disposed at the site of the `goto`. Variables not yet declared will **not** be disposed.
+```c#
+label1:
+    {
+        using var x = ...;
+        goto label1; // dispose x
+        using var y = ...; 
+    }
+```
+
+It is always permissible to jump to the same scope, when the `goto` and target `label` do not lie either side of a using declaration.
+```c#
+    goto label1; // ok, not jumping over a using declaration
+label1:
+    using var x = ...;
+label2:
+    goto label2; // ok, not jumping over a using declaration
+```
+
+
+### Other restrictions on use
+
+A using declaration may **not** appear directly inside of a case label. It may instead be used within a block inside of a case label
+
+```c#
+switch (...)
+{
+    case ...: 
+        using var x = ...; // error CS8647: A using variable cannot be used directly within a switch section (consider using braces). 
+        break;
+    
+    case ...:
+        {
+            using var y = ...; // ok
+
+            // Dispose y
+        }
+        break;
+}
+```
+
+A using declaration may **not** appear directly as part of an `out` variable declaration. It is easy to emulate this behavior however, by adding a using declaration immediately after the `out` variable:
+```c#
+if(TryGetDisposable(out var x))
+{
+    using var y = x;
+
+    // Dispose y, and thus x
+}
+```
+
+## Pattern based asynchronous disposal 
+
+With pattern based asynchronous disposal a type can be used in an `await using` statement without needing to explicitly implement `IAsyncDisposable` if it meets certain structural requirements. Specifically: 
+
+    "A reachable, Task-like returning instance method called DisposeAsync, that can be called with zero explicit arguments"
```

Comment:
>Task-like [](start = 18, length = 9)

I'm not sure where to mention this, but it should be a non-generic Task-like (`Task` rather than `Task<something>`).